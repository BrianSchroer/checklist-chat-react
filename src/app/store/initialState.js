import * as modalDialogType from '../modalDialogType';

export default {
    rooms: [],
    currentRoomId: null,
    ajaxCallsInProgressCount: 0,
    modalDialog: modalDialogType.NONE
};
