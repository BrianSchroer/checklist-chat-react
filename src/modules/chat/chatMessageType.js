export const CHAT = 'Chat';
export const ACTION = 'Action';
