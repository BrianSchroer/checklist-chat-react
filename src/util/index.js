import format from './format';
import uiHelper from './uiHelper';
import validationHelper from './validationHelper';

export {
    format as format,
    uiHelper as uiHelper,
    validationHelper as validationHelper
};
