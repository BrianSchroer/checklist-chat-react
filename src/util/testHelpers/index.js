import {shallow, mount} from 'enzyme';
import enzymeHelper from './enzymeHelper';

export {
    shallow as shallow,
    mount as mount,
    enzymeHelper as enzymeHelper
};
