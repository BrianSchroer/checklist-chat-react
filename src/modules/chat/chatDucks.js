// Chat actions/reducer "ducks" file (https://github.com/erikras/ducks-modular-redux)

import initialState from '../../app/store/initialState';
import {beginAjaxCall, ajaxCallError} from '../../app/ajaxStatusDucks';
import * as mockSignalR from '../../api/mockSignalR';

const prefix = 'checklist-chat/chat/';

// won't be used with SignalR?:
export const LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS = `${prefix}LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS`;

export const DISPLAY_CHAT_MESSAGE = `${prefix}DISPLAY_CHAT_MESSAGE`;
export const DISPLAY_CHAT_USER_ACTION = `${prefix}DISPLAY_CHAT_USER_ACTION`;
export const DISPLAY_PRIORITY_NOTIFICATION = `${prefix}DISPLAY_PRIORITY_NOTIFICATION`;
export const UPDATE_CHAT_PARTICIPANTS = `${prefix}UPDATE_CHAT_PARTICIPANTS`;

// Actions:

export const loadChatMessagesForRoomSuccess =
    (chatMessages) => ({type: LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS, chatMessages});

export const updateChatParticipants =
    (chatParticipants) => ({type: UPDATE_CHAT_PARTICIPANTS, chatParticipants});

const updateMessagesAndParticipants = (chatMessages, dispatch) => {
    dispatch(loadChatMessagesForRoomSuccess(chatMessages));

    const sortedNames = chatMessages.map(m => m.userName).sort();
    const uniqueNames = [...new Set(sortedNames)];

    const chatParticipants = uniqueNames.map(userName => ({
        id: userName,
        name: userName,
        department: 'Department',
        title: 'Title',
        connection: 'Connection'
    }));

    dispatch(updateChatParticipants(chatParticipants));
};

export function loadChatMessagesForRoom(roomId) {
    if (roomId) {
        return dispatch => {
            dispatch(beginAjaxCall());

            return mockSignalR.getChatMessages(roomId).then(chatMessages =>
                updateMessagesAndParticipants(chatMessages, dispatch)
            ).catch(error => {
                dispatch(ajaxCallError(error));
                throw (error);
            });
        };
    } else {
        return [];
    }
}

export function saveChatMessage(chatMessage, roomId) {
    return dispatch => {
        dispatch(beginAjaxCall());

        return mockSignalR.chat(roomId, chatMessage).then(() =>
        {
            mockSignalR.getChatMessages(roomId).then(chatMessages =>
                updateMessagesAndParticipants(chatMessages, dispatch));
        }).catch(error => {
            dispatch(ajaxCallError(error));
            throw (error);
        });
    };
}

// Reducers:

export function chatMessagesReducer(chatMessages = initialState.chatMessages, action) {
    const actionType = action.type;

    switch (actionType) {

        case LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS:
            return action.chatMessages;

        case DISPLAY_CHAT_MESSAGE:
        case DISPLAY_CHAT_USER_ACTION:
            return [...chatMessages, Object.assign({}, action.chatMessage)];

        default:
            return chatMessages;
    }
}

export function chatParticipantsReducer(chatParticipants = initialState.chatParticipants, action) {
    const actionType = action.type;

    switch (actionType) {

        case UPDATE_CHAT_PARTICIPANTS:
            return action.chatParticipants;

        default:
            return chatParticipants;
    }
}
