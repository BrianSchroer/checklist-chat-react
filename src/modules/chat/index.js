import * as chatMessageType from './chatMessageType';
import ChatMessageMarkdown from './components/ChatMessageMarkdown';

export {
    chatMessageType as chatMessageType,
    ChatMessageMarkdown as ChatMessageMarkdown
};
