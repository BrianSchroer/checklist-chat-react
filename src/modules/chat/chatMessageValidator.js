import {validationHelper} from '../../util';

export function validate(chatMessage) {
    return validationHelper.validateFields([{
        fieldName: 'text',
        predicate: validationHelper.missingRequiredValue(chatMessage.text),
        errorMessage: 'Message is required.'
    }]);
}
