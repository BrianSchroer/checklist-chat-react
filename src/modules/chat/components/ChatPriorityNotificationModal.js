import React, {PropTypes} from 'react';
import {connect} from 'react-redux';
import {Modal} from '../../../modules/core';
import ChatMessage from './ChatMessage';

export class ChatPriorityNotificationModal extends React.Component {

    static get propTypes()  {
        return {
            chatMessage: PropTypes.object.isRequired,
            onCloseRequest: PropTypes.func.isRequired
        };
    }

   constructor(props, context) {
        super(props, context);
    }

    render() {
        const {chatMessage, onCloseRequest} = this.props;

        const buttons = (
            <input type="button" value="Close" className="btn btn-primary"
                onClick={onCloseRequest}/>
        );

        return (
            <Modal title={`Important message from ${chatMessage.userName}`}
                onCloseRequest={onCloseRequest}
                buttons={buttons}>

                <ChatMessage chatMessage={chatMessage} />

            </Modal>
        );
    }
}

export function mapStateToProps(state, ownProps) {
    const [chatMessage] = state.modalDialogRequest.keys;
    const onCloseRequest = ownProps.onCloseRequest;

    return {chatMessage, onCloseRequest};
}

export default connect(mapStateToProps)(ChatPriorityNotificationModal);
