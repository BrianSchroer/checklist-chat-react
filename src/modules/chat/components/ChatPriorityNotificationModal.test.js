import React from 'react';
import initialState from '../../../app/store/initialState';
import {snapshotHelper} from '../../../util/testHelpers';
import {ChatPriorityNotificationModal, mapStateToProps} from './ChatPriorityNotificationModal';
import {chatMessageType} from '../../chat';

const defaultProps = {
    chatMessage: {
        timeStamp: '2016-12-08T14:57:10.222Z',
        userName: 'test userName',
        text: 'test text',
        chatMessageType: chatMessageType.CHAT
    },
    onCloseRequest: () => {}
};

const defaultOwnProps = {
    onCloseRequest: () => {}
};

function overrideProps(propOverrides) {
    return Object.assign({}, defaultProps, propOverrides);
}

function assertSnapshotMatch(propOverrides = {}) {
    snapshotHelper.assertMatch(<ChatPriorityNotificationModal {...overrideProps(propOverrides)} />);
}

function callMapStateToProps(stateOverrides, ownPropsOverrides) {
    const state = Object.assign({}, initialState, stateOverrides || {});
    const ownProps = Object.assign({}, defaultOwnProps, ownPropsOverrides || {});

    return mapStateToProps(state, ownProps);
}

describe('ChatPriorityNotificationModal', () => {
    it('should render correctly', () => {
        assertSnapshotMatch();
    });

    describe('mapStateToProps', () => {
        it('should return expected props', () => {
            const testMessage = { id: 123 };
            const stateOverrides = { modalDialogRequest: { keys: [testMessage]} };
            const props = callMapStateToProps(stateOverrides);

            expect(props.chatMessage).toBe(testMessage);
            expect(props.onCloseRequest).toBe(defaultOwnProps.onCloseRequest);
        });
    });
});
