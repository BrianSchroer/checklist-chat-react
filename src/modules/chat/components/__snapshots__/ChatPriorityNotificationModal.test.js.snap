// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatPriorityNotificationModal should render correctly 1`] = `
<div
  className="checklist-chat-modal-backdrop"
  role="dialog"
  tabIndex="-1"
>
  <div
    className="modal-dialog checklist-chat-modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="modal-header bg-primary text-white"
        >
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
          <h4
            className="modal-title display-linebreaks"
          >
            Important message from test userName
          </h4>
        </div>
        <div
          className="modal-body"
        >
          <div
            className="chat-message"
          >
            <div
              className="timestamp chat-message-timestamp"
            >
              8:57 am
            </div>
            <strong>
              test userName: 
            </strong>
            <span
              className="chat-message-text"
            >
              test text
            </span>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <input
            className="btn btn-primary"
            onClick={[Function]}
            type="button"
            value="Close"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
